### PR TITLE
feat: Breeze-aware AI agent skills system (GSoC 2026 PoC)

### LINKED ISSUE #62500: [GSOC 2026] Airflow Contribution & Verification Agent Skills
> [!IMPORTANT]
> 1. We had already decided to use contributor-docs as source of truth here instead of solely based on Breeze CLI: https://github.com/apache/airflow/issues/62500#issuecomment-4028191936e
> 2. Please add the link to your PoC (draft PR, feature-branch link on your fork, etc.) if you have **at the very beginning of your proposal** so that we don’t overlook it.

### Background

Apache Airflow’s **Breeze** environment is the de facto way to reproduce CI, run tests, and verify changes locally. It encapsulates complex tooling (Docker, integrations, static checks, tests, system verification) behind a single, consistent developer interface.

However, modern AI coding tools (e.g. Claude Code, Gemini CLI, GitHub Copilot–style agents) currently treat Airflow’s repo like any generic Python project. They **rarely**:

- Understand whether they are running inside or outside Breeze.
- Choose the correct commands for host vs. container.
- Follow the same workflows that Airflow contributors actually use (e.g. `prek`, `breeze shell`, `breeze start-airflow`).

We already expose some information through docs (e.g. `AGENTS.md`), but this mostly inflates the context window rather than giving agents a structured, machine-usable interface to Breeze.

This project aims to bridge that gap by creating an **“Airflow Breeze Contribution / Contribution Verification” AI skill** (final name TBD) that systematically encodes common contribution workflows and makes them reliably executable and testable by AI agents.

### Goal

The overarching goal is to make AI tools:

**Breeze-aware**: able to detect whether they are running inside or outside Breeze and act accordingly.
In practice, this means that for a typical contributor PR, an AI agent can:

- Run the right static checks.
- Run the right subset of tests in Breeze.
- Spin up Airflow and verify system behavior for a Dag representing the change (nice-to-have).
- **Do all of the above while respecting host/container boundaries**.

Additionally, the solution should be **consistency-focused**, meaning that we want to **keep Breeze CLI as the single source of truth for agent skills**. This can be achieved by auto-syncing CLI docstrings and behaviors into the AI skill using existing tooling (e.g. `prek`), ensuring that the skill definitions always reflect the current state of the Breeze CLI.

##### 1. Environment Awareness & Detection

- Design and implement a simple, robust mechanism for the agent skills to detect:
  - “Host” vs “inside Breeze container”.
  - Relevant environment variables, markers, or file paths that indicate context.
- Encode decision logic for when to run:
  - Host-only commands (e.g. `breeze shell`, `breeze start-airflow`, `git` operations).
  - Container-only commands (e.g. `pytest`, `airflow ...`).
- Provide a clear API/contract that AI tools can call to query current context and get recommended commands.

**Note**: Maybe we need to add some explicit markers, files in the repo, or write a small helper script that can be called to determine context in a reliable way. Or maybe we can rely on existing environment variables or filesystem cues. This is an open design question to explore.

##### 2. Modeling Core Contributor Workflows as Skills

Based on the three scenarios described, define and implement skills that represent common contribution flows:

**Scenario 1: Static checks pass**

- Stage changes (`git add ...`).
- Run `prek`.
- Collect and surface failures in a structured way so that an agent can fix them.

**Scenario 2: Unit tests in Breeze**

- Start or attach to a Breeze container with `breeze shell` or `breeze exec`.
- Run `pytest` with a targeted module/test path (not the whole suite).
- Then the agent can inspect results and decide on next steps (e.g. fix code, exit Breeze).

##### 3. Syncing with Breeze CLI as Source of Truth (via `prek`)

- Investigate existing Breeze CLI docstrings and structure.
- Define a mapping from Breeze commands (and their docstrings) to skill definitions, paths, and parameters.
- Implement a `prek` hook that:
  - Generates or updates the agent skills definition files from Breeze CLI docstrings.
  - Fails when drift is detected (e.g. a command changed but the skill spec was not updated).
- Integrate these checks into existing static check pipelines so the skills stay in sync automatically.

##### 4. Evaluation & Test Harness

- Design a testable user scenario or “exam” that simulates a typical contribution workflow (e.g. fixing a simple bug, adding a small feature) to verify that the added skills work as intended.
- Add unit tests for any additional scripts or helper functions created.

##### 5. Documentation & Developer Guide

- Add or extend documentation (e.g. `AGENTS.md`, Breeze docs) to:
  - Describe the new Breeze-aware skills.
  - Show example workflows for human contributors and AI tools.
  - Document how other tools can integrate with the skills (e.g. path to spec file, key commands).

### Advanced Tasks (Optional / Stretch Goals)

**Scenario: System behavior verification**

- Write a Dag representing the feature/bugfix being contributed (or use an existing one).
- Run `breeze start-airflow` (with `--integration` when needed).
- Trigger the Dag via CLI (instead of UI) and wait for completion.
- Inspect logs/status to determine success/failure from the TaskInstance logs.
- Inspect logs/status from all the component services (scheduler, api-server, triggerer, etc) to determine if there are any underlying issues.
- The agent can then decide to fix code, fix the Dag, or exit Breeze based on the results.

### Expected Outcome

By the end of the project, we expect:

- A **Breeze-aware AI skill** that can:
  - Detect host vs. container context.
  - Choose appropriate commands and environment transitions.
- The AI toolings will be "smart-enough" to handle the core workflows for contributions, including:
  - Static checks with `prek`.
  - Targeted unit tests in Breeze.
  - Continue iterating based on results (e.g. fix code, fix tests, exit).
- A **sync mechanism** (likely using `prek`) that:
  - Keeps Breeze CLI and the skill definitions in sync.
  - Fails CI when they diverge, ensuring Breeze remains the single source of truth.
- Initial **evaluation “exam(s)”** and test harnesses that:
  - Verify that an implementation of the skill behaves correctly on at least the core scenarios.
- Updated **documentation** explaining how contributors and AI tools can make use of the new capability.

A successful project will make it much easier for future AI tooling (IDEs, CLIs, bots) to interact with Breeze in a reliable and Airflow-native way, increasing contributor productivity and lowering the barrier to entry.

### Recommended Skills

- **Programming & Tooling**
  - Solid Python skills (CLI tools, packaging, basic testing).
  - Familiarity with Docker and containerized development environments.
  - Experience with writing or using CLIs and handling subprocesses.
- **Dev Workflow & CI**
  - Understanding of typical open source contribution workflows (git, PRs, static checks, unit tests, pre-commit).
  - Exposure to CI systems and concepts of reproducible environments.
- **AI/Agents**
  - Interest in or experience with AI coding assistants, Agent Skills, tool-calling, or agent frameworks.
  - Comfort reasoning about what “smart enough” means in terms of concrete, testable behaviors.
- **Airflow/Breeze (Nice to Have)**
  - Basic knowledge of Apache Airflow concepts (Dags, tasks, operators).
  - Prior use of Breeze for development or testing is a plus, but not strictly required.

Motivation to work at the intersection of developer experience, tooling, and AI is more important than prior deep expertise in all of these areas.

### Mentors

- Jason Liu (GitHub: @jason810496, Slack: Zhe-You(Jason) Liu)
- Jarek Potiuk (GitHub: @potiuk, Slack: Jarek Potiuk)
- [`#gsoc` Slack Channel in Apache Airflow workspace](https://apache-airflow.slack.com/archives/CSC0FLNJF)

### Learning Materials

- Airflow Breeze documentation: https://github.com/apache/airflow/blob/main/dev/breeze/doc/README.rst
- Recent Airflow Dev Mailing List discussion regarding Agent Skills/ Agents:
    - [[CONTRIBUTION OPPORTUNITY] Help Define Translation Agent Skills for Airflow Terminology](https://lists.apache.org/thread/r6m7wkkflpqzjjvcmg93xwlhn6nqjylt)
    - [New: AGENTS.md overhaul + Copilot catches AI-slop PRs automatically](https://lists.apache.org/thread/szygtjs2t7mb6c2kk4zjygslxf0d4od8)
- Airflow `prek` (`pre-commit`) hooks entrypoint: https://github.com/apache/airflow/blob/main/.pre-commit-config.yaml
- Modern Python monorepo for Apache Airflow (by Jarek): https://medium.com/apache-airflow/modern-python-monorepo-for-apache-airflow-part-1-1fe84863e1e1
- `pre-commit`: https://pre-commit.com/
- `prek`: https://github.com/j178/prek

### Committer

- [x] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by Subham-KRLX**
Problem

AI coding tools like Claude Code, Gemini CLI, and Cursor treat Apache Airflow as a generic Python repository. They don't understand the boundary between the host OS and the Breeze container which causes real failures in contributor workflows.

Running pytest on the host fails because PostgreSQL binaries aren't there. Running git commit inside Breeze fails because .git is not mounted. When commands fail silently agents start hallucinating fixes for bugs that don't exist.

Solution

This PR adds a machine-readable library of contributor workflows following the agentskills.io specification. It gives AI tools a clear map of what to run where to run it and what to avoid.

breeze_context.py detects whether code is running on the host or inside Breeze using a simple chain: BREEZE_HOME → .dockerenv → /opt/airflow. It also blocks host-only operations like git add from running inside the container.

.agents/skills/ contains 3 skills covering testing, staging, and linting. Each skill tells the agent exactly which command to use depending on context.

validate_skills.py is a pre-commit hook that checks every skill file follows the correct structure. It handles Airflow's SPDX license headers so the two systems don't conflict.

AGENTS.md → CONSTITUTION.md → .agents/skills/ forms a three-level hierarchy: short entrypoint, full rules, then context-aware execution.

The norecursedirs fix in pyproject.toml, airflow-ctl, and task-sdk resolves a SystemExit: 1 test collection error caused by pytest scanning dev/ directories it shouldn't touch.

Verification Proof:

✅ 14/14 scripts tests passing (context detection, command routing, guardrails, edge cases)
✅ 15 files changed all PoC-specific no unrelated changes
✅ ruff format and ruff check clean on all new Python files
✅ validate_skills.py all 3 skills conform to agentskills.io standard

closes: #62500

Was generative AI tooling used to co-author this PR? 
Yes — Claude(Code Research and PR Sanitization)

### CODE REVIEW

File: scripts/ci/prek/breeze_context.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+        # Define all available skills
+        skills: dict[str, Any] = {
+            "stage-changes": {
+                # git add operates on the host working tree — not inside the container.
+                # This is explicitly a host-only operation per the contributor workflow.
+                "host": "git add {path}",
+                "breeze": None,
+                "preferred": "host",
+                "params": {"path": {"required": True}},
```
[the remark this answers is not in the document]
Comment (Subham-KRLX) [reply scripts/ci/prek/breeze_context.py:176]: Fixed in 721fc67: updated to uv run --project airflow-core ... with aligned tests/docs resolved on latest diff.

**Comment by Subham-KRLX** [comment]
Updated CONSTITUTION.md to include the full Commands section that was missing it now contains all uv run, breeze testing, breeze run, and prek commands alongside the host/Breeze boundary table this completes the AGENTS.md (short entrypoint) → CONSTITUTION.md (full rules + commands) → .agents/skills/ (context-aware execution) hierarchy as described in the proposal.

**Comment by Subham-KRLX** [comment]
The recurring AGENTS.md conflict is expected this PR adds the .agents/skills/ section to it and main keeps expanding the same file resolution is always to keep this PR's short entrypoint version. Happy to re-sync as needed.

File: airflow-ctl/pyproject.toml [hunk 2]
```diff
@@ -199,6 +199,7 @@ markers = [
 ]
 norecursedirs = [
     ".eggs",
+    "dev",
```
Comment (potiuk) [inline airflow-ctl/pyproject.toml]: Why this?

File: chart/files/statsd-mappings.yml [hunk 3]
```diff
@@ -92,7 +92,6 @@ mappings:
     name: "airflow_pool_used_slots"
     labels:
       pool: "$1"
-
   - match: airflow.pool.starving_tasks.*
```
Comment (potiuk) [inline chart/files/statsd-mappings.yml:96]: Why removed line here

File: scripts/ci/prek/breeze_context.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+Breeze context detection and command selection.
+
+This module exists for agent workflows that must reliably choose between:
+
+- "host" execution (commands intended to run on the contributor machine)
+- "breeze" execution (commands intended to run inside the Breeze container)
+
+Architecture
+------------
+Agents call :meth:`BreezieContext.get_command` with a skill id and parameters.
```
Comment (potiuk) [inline scripts/ci/prek/breeze_context.py]: We do not use :meth: usually and `BreezieContext` has spelling error - so it's for sure not used . Can you fix it please?

File: scripts/ci/prek/breeze_context.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+            skill_id,
+            context,
+            kwargs,
+            command,
+        )
+        return command
+
+
+# Backwards compatibility alias
+BreezieContext = BreezeContext
```
Comment (potiuk) [inline scripts/ci/prek/breeze_context.py]: Why Breezie ? Why compatibility?

File: pyproject.toml [hunk 6]
```diff
@@ -946,6 +946,7 @@ norecursedirs = [
     "airflow-core/tests/unit/dags_with_system_exit",
     "airflow-core/tests/unit/dags_corrupted",
     "airflow-core/tests/unit/dags",
+    "dev",
```
Comment (potiuk) [inline pyproject.toml]: Why this?

File: .agents/skills/run-unit-tests/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+2. run-static-checks
+3. → **run-unit-tests** (this skill)
+
+Run static checks first to avoid failing tests due to trivial formatting issues.
+
+Prerequisites
+-------------
+
+- **Host:** `uv` must be installed and the project synced.
+- **Breeze:** Docker must be running and the Breeze image built. Start a Breeze
```
Comment (potiuk) [inline .agents/skills/run-unit-tests/SKILL.md]: I think there is no need to run shell as prerequisite.

Also I think it would be great to add that in case breeze image needs rebuilding, it shoudl be rebuilt (answer `yes` to the question of rebuilding).

File: .agents/skills/stage-changes/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+Host-Only Guardrail
+-------------------
+
+This skill operates on the **host** working tree. Running `git` commands inside a
+Breeze container will silently operate on container-local paths that differ from your
+actual working tree.
+
+**Before running:** Confirm you are on the host by checking:
+
+- `/.dockerenv` does **not** exist
```
Comment (potiuk) [inline .agents/skills/stage-changes/SKILL.md]: I am not sure those guardrails here are really needed - and I think they are just somewhat wrong

BREEZE="true" should be all that is needed. I am not sure where `BREEZE_HOME` is from - it seems only added in tests here and is not actually set in any other place, so I am not sure how it would work ?

Comment (potiuk) [reply .agents/skills/stage-changes/SKILL.md]: Also I think this all should be tested with `devcontainer` setup. In devcontainer, we use `breeze ci` built  image - but by definition the devcontainers run inside container - so likely there is no good way to exit from it ?  I am not sure if agents work with devcontainers and how - but I think this scenario should be tested what if someone uses devcontainers (say codespaces, but also VSCode and IntelliJ have built-in devcontainer support and some people might opt to choose it. 

I believe in Dev container setup, all the sources - including .git are mounted to the container - and genereally speaking `git` should work there. however of course devil is in details - things like authentication and (especially nowadays with agents and `claude -w` or generally using. worktrees (which finally got used a lot thanks to agentic workflows) - I am not sure how it works with devcontainers.

It shoudl be definitely checked and ideally we should bave more complete scenarios describing the different workflows - including worktrees and devcontainers.

File: AGENTS.md [hunk 9]
```diff
@@ -1,392 +1,56 @@
- <!-- SPDX-License-Identifier: Apache-2.0
+<!-- SPDX-License-Identifier: Apache-2.0
       https://www.apache.org/licenses/LICENSE-2.0 -->
 
 # AGENTS instructions
 
-## Environment Setup
+## Purpose
+
+This file is a short map for agents. Load this first, then load CONSTITUTION.md for
```
Comment (potiuk) [inline AGENTS.md:8]: I like this.

File: scripts/ci/prek/breeze_context.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (potiuk) [inline scripts/ci/prek/breeze_context.py:1]: Learning more about agents - I am not sure if we **need** that - also I think detection of BREEZE should be far easier - just presence of `BREEZE` env variable should be enough I think - with the devcontainer case handled in a special way I guess - also we have full control over what variables are set in BREEZE and devcontainer - and in case we need it - we can very easily add new variables, to differentiate. I am not sure why we would have to run this python script to detect breeze. I guess something like that would work reliably:

> To detect if you are in breeze check if `BREEZE` variable is set

That should be quite enough for agent to detect where it is.

Or am I missing something ? Why would we complicate it like that?

**Comment by potiuk** [comment]
Comments added - also there are conflicts to solve.

File: airflow-ctl/pyproject.toml (hunk 2 above)
Comment (Subham-KRLX) [reply airflow-ctl/pyproject.toml]: I have removed those changes and synced the files with main in the latest squash

File: chart/files/statsd-mappings.yml (hunk 3 above)
Comment (Subham-KRLX) [reply chart/files/statsd-mappings.yml:96]: I accidentally nuked that blank line during a rebase but it's back to normal now.

File: scripts/ci/prek/breeze_context.py (hunk 4 above)
Comment (Subham-KRLX) [reply scripts/ci/prek/breeze_context.py]: Fixed the typo and removed the :meth: formatting. Good call!

File: scripts/ci/prek/breeze_context.py (hunk 5 above)
Comment (Subham-KRLX) [reply scripts/ci/prek/breeze_context.py]: That was a leftover from an earlier draft I have removed the alias entirely.

File: pyproject.toml (hunk 6 above)
Comment (Subham-KRLX) [reply pyproject.toml]: Same as above removed the unrelated noise and synced with main

File: .agents/skills/run-unit-tests/SKILL.md (hunk 7 above)
Comment (Subham-KRLX) [reply .agents/skills/run-unit-tests/SKILL.md]: Agreed.

File: .agents/skills/stage-changes/SKILL.md (hunk 8 above)
Comment (Subham-KRLX) [reply .agents/skills/stage-changes/SKILL.md]: I have simplified the guardrails to just check for BREEZE="true" and added a new section to CONSTITUTION.md specifically for DevContainers and Worktrees to clear up the different workflow scenarios.

File: scripts/ci/prek/breeze_context.py (hunk 10 above)
Comment (Subham-KRLX) [reply scripts/ci/prek/breeze_context.py:1]: Yeah checking the env var is definitely easier i kept the script because it acts as a 'router' it handles the mapping so agents don't have to worry about the if/else logic for every command they just ask for a skill and get the right string back instantly.

**Comment by potiuk** [comment]
@Subham-KRLX A few things need addressing before review — see our [Pull Request quality criteria](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-quality-criteria).

- :x: **Pre-commit / static checks**. See [docs](https://github.com/apache/airflow/blob/main/contributing-docs/08_static_code_checks.rst).

No rush.

---

_Note: This comment was drafted by an AI-assisted triage tool and may contain mistakes. Once you have addressed the points above, an Apache Airflow maintainer — a real person — will take the next look at your PR. We use this [two-stage triage process](https://github.com/apache/airflow/blob/main/contributing-docs/25_maintainer_pr_triage.md#why-the-first-pass-is-automated) so that our maintainers' limited time is spent where it matters most: the conversation with you._

---
Drafted-by: Claude Code (Opus 4.7); reviewed by @potiuk before posting

**Comment by potiuk** [comment]
@Subham-KRLX — I've removed the `ready for maintainer review` label because the next step here is yours: this branch now has **merge conflicts with `main`**. It'll go back into the maintainer queue automatically once you rebase onto the latest `main` and push — no need to re-add the label by hand. There's no rush.

<sub>Automated triage note drafted by an AI-assisted tool — may get things wrong; a real Apache Airflow maintainer takes the next look once it's green. ([why automated](https://github.com/apache/airflow/blob/main/contributing-docs/25_maintainer_pr_triage.md))</sub>

---
Drafted-by: Claude Code (Opus 4.8); reviewed by @potiuk before posting

**Comment by potiuk** [comment]
@Subham-KRLX This draft PR has been inactive for 46 days since the last triage note and no response from the author. Closing to keep the queue clean.

You are welcome to reopen this PR when you resume work, or to open a new one addressing the issues previously raised. There is no rush — take your time.

---

_Note: This comment was drafted by an AI-assisted triage tool and may contain mistakes. If you have questions, an Apache Airflow maintainer — a real person — is happy to help. We use this [two-stage triage process](https://github.com/apache/airflow/blob/main/contributing-docs/25_maintainer_pr_triage.md#why-the-first-pass-is-automated) so that our maintainers' limited time is spent where it matters most: the conversation with you._
